src/main/java/org/cryptomator/cryptofs/health/type/CiphertextFileTypeCheck.java

```diff
@@ -0,0 +1,160 @@
+import org.cryptomator.cryptofs.health.api.HealthCheck;
+import org.cryptomator.cryptolib.api.Cryptor;
+import org.cryptomator.cryptolib.api.Masterkey;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.EnumSet;
+import java.util.Set;
+import java.util.function.Consumer;
+
+/**
+ * Checks for each c9r or c9s dir in the vault structure if its {@link org.cryptomator.cryptofs.common.CiphertextFileType} can be determined.
+ * <p>
+ * The type is based on the presence of a signature file.
+ * The signature files are {@value org.cryptomator.cryptofs.common.Constants#DIR_FILE_NAME}, {@value org.cryptomator.cryptofs.common.Constants#SYMLINK_FILE_NAME} and {@value org.cryptomator.cryptofs.common.Constants#CONTENTS_FILE_NAME}.
+ * Note: For c9r dirs, only the dir and symlink sig files are tested.
+ */
+public class CiphertextFileTypeCheck implements HealthCheck {
+
+	private static final Logger LOG = LoggerFactory.getLogger(CiphertextFileTypeCheck.class);
+	private static final int MAX_TRAVERSAL_DEPTH = 3;
+
+	@Override
+	public String name() {
+		return "Resource Type Check";
+	}
+
+	@Override
+	public void check(Path pathToVault, VaultConfig config, Masterkey masterkey, Cryptor cryptor, Consumer<DiagnosticResult> resultCollector) {
+
+		// scan vault structure:
+		var dataDirPath = pathToVault.resolve(Constants.DATA_DIR_NAME);
+		var dirVisitor = new CiphertextFileTypeCheck.DirVisitor(resultCollector);
+		try {
+			Files.walkFileTree(dataDirPath, Set.of(), MAX_TRAVERSAL_DEPTH, dirVisitor);
+		} catch (IOException e) {
+			LOG.error("Traversal of data dir failed.", e);
+			resultCollector.accept(new CheckFailed("Traversal of data dir failed. See log for details."));
+			return;
+		}
+	}
+
+	// visible for testing
+	static class DirVisitor extends SimpleFileVisitor<Path> {
+
+		private final Consumer<DiagnosticResult> resultCollector;
+
+		public DirVisitor(Consumer<DiagnosticResult> resultCollector) {this.resultCollector = resultCollector;}
+
+		@Override
+		public FileVisitResult visitFile(Path dir, BasicFileAttributes attrs) {
+			switch (determineFileType(dir, attrs)) {
+				case C9R_DIR -> checkCiphertextTypeC9r(dir);
+				case C9S_DIR -> checkCiphertextTypeC9s(dir);
+				case UNRELATED -> {}
+			}
+			;
+			return FileVisitResult.CONTINUE;
+		}
+
+		DirType determineFileType(Path p, BasicFileAttributes attrs) {
+			var dirName = p.getFileName().toString();
+			boolean isDir = attrs.isDirectory();
+
+			if (isDir && dirName.endsWith(Constants.CRYPTOMATOR_FILE_SUFFIX)) {
+				return DirType.C9R_DIR;
+			} else if (isDir && dirName.endsWith(Constants.DEFLATED_FILE_SUFFIX)) {
+				return DirType.C9S_DIR;
+			} else {
+				return DirType.UNRELATED;
+			}
+		}
```

Comment:
if you only check for `.c9r` and `.c9s` dirs (not _files_), you can simplify this by using `preVisitDirectory(...)` instead

Reply:
That is true, (before f138d82 it was the case)  but then i need to increase `MAX_DEPTH` to 4 and should return `SKIP_SUBTREE` if a c9r/c9s dir is detected. Hence, the "complexity" will be just shifted elsewhere. Also, with an increased MAX_DEPTH we could visit more _unrelated_ files/directories.

Also performant wise there is nearly no gain: Using only the `preVisitDirectory(...)` method, the `visitFile(...)` has less no branching (due to SimpleFileVisitor) and under the assumption there are always more files than directories to visit, it should be faster. But the time-consuming task is reading the attributes and we do not skip any files or directories,

Reply:
ok. btw I didn't know that `visitFile` is invoked for directories as well

Reply:
> ok. btw I didn't know that `visitFile` is invoked for directories as well

Well yes and no. Directories are only visited at maxDepth, see [Files.walkFileTree](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/nio/file/Files.html#walkFileTree(java.nio.file.Path,java.util.Set,int,java.nio.file.FileVisitor)), third last paragraph.